File: api/src/main/java/org/apache/gravitino/rel/expressions/sorts/SortOrders.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.gravitino.rel.expressions.sorts;

import java.util.Objects;
import org.apache.gravitino.rel.expressions.Expression;

/** Helper methods to create SortOrders to pass into Apache Gravitino. */
public class SortOrders {

  /** NONE is used to indicate that there is no sort order. */
  public static final SortOrder[] NONE = new SortOrder[0];

  /**
   * Create a sort order by the given expression with the ascending sort direction and nulls first
   * ordering.
   *
   * @param expression The expression to sort by
   * @return The created sort order
   */
  public static SortImpl ascending(Expression expression) {
    return of(expression, SortDirection.ASCENDING);
  }

  /**
   * Create a sort order by the given expression with the descending sort direction and nulls last
   * ordering.
   *
   * @param expression The expression to sort by
   * @return The created sort order
   */
  public static SortImpl descending(Expression expression) {
    return of(expression, SortDirection.DESCENDING);
  }

  /**
   * Create a sort order by the given expression with the given sort direction and default null
   * ordering.
   *
   * @param expression The expression to sort by
   * @param direction The sort direction
   * @return The created sort order
   */
  public static SortImpl of(Expression expression, SortDirection direction) {
    return of(expression, direction, direction.defaultNullOrdering());
  }

  /**
   * Create a sort order by the given expression with the given sort direction and null ordering.
   *
   * @param expression The expression to sort by
   * @param direction The sort direction
   * @param nullOrdering The null ordering
   * @return The created sort order
   */
  public static SortImpl of(
      Expression expression, SortDirection direction, NullOrdering nullOrdering) {
    return new SortImpl(expression, direction, nullOrdering);
  }

  /**
   * Create a sort order by the given expression with the given sort direction and null ordering.
   */
  public static final class SortImpl implements SortOrder {
    private final Expression expression;
    private final SortDirection direction;
    private final NullOrdering nullOrdering;

    private SortImpl(Expression expression, SortDirection direction, NullOrdering nullOrdering) {
      this.expression = expression;
      this.direction = direction;
      this.nullOrdering = nullOrdering;
    }

    @Override
    public Expression expression() {
      return expression;
    }

    @Override
    public SortDirection direction() {
      return direction;
    }

    @Override
    public NullOrdering nullOrdering() {
      return nullOrdering;
    }

    @Override
    public boolean equals(Object o) {
      if (this == o) {
        return true;
      }
      if (o == null || getClass() != o.getClass()) {
        return false;
      }
      SortImpl sort = (SortImpl) o;
      return Objects.equals(expression, sort.expression)
          && direction == sort.direction
          && nullOrdering == sort.nullOrdering;
    }

    @Override
    public int hashCode() {
      return Objects.hash(expression, direction, nullOrdering);
    }

    @Override
    public String toString() {
      return "SortImpl{"
          + "expression="
          + expression
          + ", direction="
          + direction
          + ", nullOrdering="
          + nullOrdering
          + '}';
    }
  }

  private SortOrders() {}
}

File: clients/client-python/gravitino/api/expressions/sorts/sort_orders.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
from typing import List

from gravitino.api.expressions.expression import Expression
from gravitino.api.expressions.sorts.null_ordering import NullOrdering
from gravitino.api.expressions.sorts.sort_direction import SortDirection
from gravitino.api.expressions.sorts.sort_order import SortOrder


class SortImpl(SortOrder):

    def __init__(
        self,
        expression: Expression,
        direction: SortDirection,
        null_ordering: NullOrdering,
    ):
        """Initialize the SortImpl object."""
        self._expression = expression
        self._direction = direction
        self._null_ordering = null_ordering

    def expression(self) -> Expression:
        return self._expression

    def direction(self) -> SortDirection:
        return self._direction

    def null_ordering(self) -> NullOrdering:
        return self._null_ordering

    def __eq__(self, other: object) -> bool:
        """Check if two SortImpl instances are equal."""
        if not isinstance(other, SortImpl):
            return False
        return (
            self.expression() == other.expression()
            and self.direction() == other.direction()
            and self.null_ordering() == other.null_ordering()
        )

    def __hash__(self) -> int:
        """Generate a hash for a SortImpl instance."""
        return hash((self.expression(), self.direction(), self.null_ordering()))

    def __str__(self) -> str:
        """Provide a string representation of the SortImpl object."""
        return (
            f"SortImpl(expression={self._expression}, "
            f"direction={self._direction}, null_ordering={self._null_ordering})"
        )


class SortOrders:
    """Helper methods to create SortOrders to pass into Apache Gravitino."""

    # NONE is used to indicate that there is no sort order
    NONE: List[SortOrder] = []

    @staticmethod
    def ascending(expression: Expression) -> SortImpl:
        """Creates a sort order with ascending direction and nulls first."""
        return SortOrders.of(expression, SortDirection.ASCENDING)

    @staticmethod
    def descending(expression: Expression) -> SortImpl:
        """Creates a sort order with descending direction and nulls last."""
        return SortOrders.of(expression, SortDirection.DESCENDING)

    @staticmethod
    def of(
        expression: Expression,
        direction: SortDirection,
        null_ordering: NullOrdering = None,
    ) -> SortImpl:
        """Creates a sort order with the given direction and optionally specified null ordering."""
        if null_ordering is None:
            null_ordering = direction.default_null_ordering()
        return SortImpl(expression, direction, null_ordering)
